### ISSUE TITLE
Can't determine if state tools button is inactive or active for Materia GTK theme

### ISSUE THREAD

**Issue by Keyikedalube** [synfig/synfig#2200]
**Describe the bug**
State tools button's focusability is making it hard to determine if a button is active or inactive for Materia GTK theme

**To Reproduce**
1. Activate rectangle tool
2. Click on any active button
3. Inactive button is still vaguely highlighted because focus is set to this button

**Expected behavior**
When button is inactive/untoggled it should appear like the rest without grabbing focus.

**Screencast**

https://user-images.githubusercontent.com/24214509/123899279-7e220000-d984-11eb-971f-8d2ab414bdfa.mp4

**System information:**
 - OS: Fedora Linux
 - OS version: 34
 - Commit ID: 4273435309266fa4181269fdb3e9cc454cf14337

**Additional comments**
On Adwaita, a border is shown instead when button state is changed from active to inactive

**Comment by mohamedAdhamc** [comment]
is this still an issue?  if it is i would be willing to work on it.

**Comment by Keyikedalube** [comment]
@mohamedAdhamc Yes this is still an issue maybe only for me and probably other materia gtk theme users.

If you'd like to crack this issue I can assign you :)

**Comment by mohamedAdhamc** [comment]
> @mohamedAdhamc Yes this is still an issue maybe only for me and probably other materia gtk theme users.
> 
> If you'd like to crack this issue I can assign you :)

yes, sure I'd like to have a go at it.

**Comment by mohamedAdhamc** [comment]
> @mohamedAdhamc Yes this is still an issue maybe only for me and probably other materia gtk theme users.
> 
> If you'd like to crack this issue I can assign you :)

Hey sorry was carried away a bit in the past 2 days with all the build and IDE configuration errors however I will be starting to try and  tackle this, but I just have a few questions to clarify things here a little.

1- What is the expected behaviour here exactly ? My understanding is that maybe you can have multiple ones active and the problem is that inactive ones are still vaguely highlighted. Or is it that only one should be active and they should be grouped in a mutually exclusive way for example.

2- Are there any specific places or functions where you would think it'd  be helpful to look at? I'm guessing I should start state_rectangle.cpp and see where that leads me.

**Comment by Keyikedalube** [comment]
> 1- What is the expected behaviour here exactly ? My understanding is that maybe you can have multiple ones active and the problem is that inactive ones are still vaguely highlighted.

Yes, that's correct :)

> I'm guessing I should start at state_rectangle.cpp and see where that leads me.

Rightly so.

**Comment by mohamedAdhamc** [comment]
> > 1- What is the expected behaviour here exactly ? My understanding is that maybe you can have multiple ones active and the problem is that inactive ones are still vaguely highlighted.
> 
> Yes, that's correct :)
> 
> > I'm guessing I should start at state_rectangle.cpp and see where that leads me.
> 
> Rightly so.

Hey, in Adwaita it seems to be working normally -no (On Adwaita, a border is shown instead when button state is changed from active to inactive )- was it already fixed for Adwaita?

**Comment by mohamedAdhamc** [comment]
Could this bug possibly be related to the Materia GTK theme itself and not Synfigs code?   or is there certainly a flaw somewhere in there but for some reason it isn't affecting Adwaita... I will be trying to figure out the answer to this question first. However, if this is something already known it'd be appreciated to know to possibly save some time.

**Comment by Keyikedalube** [comment]
> this would be a theme specific problem

Correct because each theme has its own color tone. Materia seems to choose a slightly higher color tone when togglable buttons are switched from active to inactive that's why their active/inactive state seems confusing as demonstrated in my screencast above.
However, this is not a theme bug because Synfig still keeps grabbing button focus even on inactive state when they should be let go instead.

> this is also not rectangle tool-specific only I have realized this also happens for other types of tools

Hence the title _"Can't determine if state tools button..."_ :)
If you figured a fix for one state tool then the rest is duplicating the same behavior for other state tools; circle, bline, etc

**Comment by Keyikedalube** [comment]
Here's a good example of [Materia theme bug](https://github.com/synfig/synfig/issues/2186).

Nothing could be done from Synfig side to rectify that.

**Comment by mohamedAdhamc** [comment]
> > this would be a theme specific problem
> 
> Correct because each theme has its own color tone. Materia seems to choose a slightly higher color tone when togglable buttons are switched from active to inactive that's why their active/inactive state seems confusing as demonstrated in my screencast above. However, this is not a theme bug because Synfig still keeps grabbing button focus even on inactive state when they should be let go instead.
> 
> > this is also not rectangle tool-specific only I have realized this also happens for other types of tools
> 
> Hence the title _"Can't determine if state tools button..."_ :) If you figured a fix for one state tool then the rest is duplicating the same behavior for other state tools; circle, bline, etc

Hey, so some buttons like in the toolbar for example the show grid button. when active it doesn't grab focus, and it seems to be working just fine. would it cause another issue if grabbing focus was just straight out deactivated for all the layer types inside the tool (for ecample the rectangle tool's layer buttons as in rectangle layer, region layer, etc..)

**Comment by mohamedAdhamc** [comment]
update: found a possible solution to remove this focus (dotted rectangle thing)
it's pretty much disabling the focus for these layer buttons only. doing so successfully stopped these buttons from grabbing focus and it doesnt seem to be affecting other things, and synfig is building correctly should I make a pull request for just the rectangle tool as a demo first ?

**Comment by ice0** [comment]
Hi guys!
Author of Materia theme says this is intended behavior (https://github.com/nana-4/materia-theme/issues/356#issuecomment-462192375).

For example: 
https://github.com/nana-4/materia-theme/issues/356
https://github.com/nana-4/materia-theme/issues/409

All issues: https://github.com/nana-4/materia-theme/issues?q=is%3Aissue+focus+is%3Aclosed

I think it's not the best idea to disable focus because we lose keyboard navigation.
Unfortunately, I can't say what we can do from our side right now.
Perhaps we can change the icons for the disabled/enabled state (eg convert them to grayscale for the disabled state)

**Comment by mohamedAdhamc** [comment]
> Hi guys! Author of Materia theme says this is intended behavior ([nana-4/materia-theme#356 (comment)](https://github.com/nana-4/materia-theme/issues/356#issuecomment-462192375)).
> 
> For example: [nana-4/materia-theme#356](https://github.com/nana-4/materia-theme/issues/356) [nana-4/materia-theme#409](https://github.com/nana-4/materia-theme/issues/409)
> 
> All issues: https://github.com/nana-4/materia-theme/issues?q=is%3Aissue+focus+is%3Aclosed
> 
> I think it's not the best idea to disable focus because we lose keyboard navigation. Unfortunately, I can't say what we can do from our side right now. Perhaps we can change the icons for the disabled/enabled state (eg convert them to grayscale for the disabled state)

how about maybe just not grabbing focus from mouse clicks as a better alternative to disabling it altogether. Also I have thought of the keyboard navigation part is it really of use there ? i.e do some user use it for layer type selection ? becuase for me it seems like it would just be an inconvenient way of selecting layers.

**Comment by Keyikedalube** [comment]
> I think it's not the best idea to disable focus because we lose keyboard navigation

From https://github.com/synfig/synfig/pull/2577#issuecomment-1068719622 I made an alternate suggestion to use button signals:

_"If we wanted to keep the keyboard tab input intact we could use signals to detect when the button is toggled and thereby grabbing/ungrabbing focus intelligently. Code implementation would be longer though."_

**Comment by Keyikedalube** [comment]
> do some user use it for layer type selection ?

Sometimes I occasionally hit tab key if I wanted to avoid mouse cursor changing combobox or spinbuttons data. But that is very very very rare.

**Comment by mohamedAdhamc** [comment]
> > do some user use it for layer type selection ?
> 
> Sometimes I occasionally hit tab key if I wanted to avoid mouse cursor wheel changing combobox or spinbuttons data. But that is very very very rare.

yes thats why I originally thought it wouldnt be such a big hassle to disable it altogether. However which do you think might be a better approach disbling grabbing focus just for mouse clicks or using signals. I think the former would be much easier and I cant think of any drawbacks except maybe that if you do the layer selection using the keyboard then we would be back at the same issue kind of.

**Comment by Keyikedalube** [comment]
Since @ice0 wants to keep keyboard navigation I guess using signals would keep both side happy ;P

**Comment by mohamedAdhamc** [comment]
> Since @ice0 wants to keep keyboard navigation I guess using signals would keep both side happy ;P

I believe disabling focus just for mouse clicks would still keep keyboard navigation. I think we have a similar style thing for the top right toobar where they can be accesed via keyboard navigation but clicking them by mouse doesnt grab focus.

**Comment by mohamedAdhamc** [comment]
@Keyikedalube  done.

**Comment by ice0** [comment]
I think this is not the best idea, because we are breaking GTK behavior. 
Moreover, if we continue along this path, we will have to change all the other components, not just this one. 
As you can notice, there are a lot of issues about visual similarity of elements in this theme, but the author just closes these issues. 
Therefore, it is better to check how other programs like Krita or Gimp solves this problem.
If they didn't so better to leave it as it is. 
Because this problem can be easily solved by the theme author by sharper color differentiation or by adding a border around the focused element.

**Comment by mohamedAdhamc** [comment]
> I think this is not the best idea, because we are breaking GTK behavior. Moreover, if we continue along this path, we will have to change all the other components, not just this one. As you can notice, there are a lot of issues about visual similarity of elements in this theme, but the author just closes these issues. Therefore, it is better to check how other programs like Krita or Gimp solves this problem. If they didn't so better to leave it as it is. Because this problem can be easily solved by the theme author by sharper color differentiation or by adding a border around the focused element.

But isn't it also a bit weird looking even in adwaita when you make the button inactive to have some sort of dotted rectangle around it. I believe not having the focus when it's inactive looks less confusing, and better visually. Is there a point here or is that just me thinking too much xd ?

**Comment by Keyikedalube** [comment]
This issue is neither Synfig's nor 3rd party theme author's fault even though I stated so in my comment above https://github.com/synfig/synfig/issues/2200#issuecomment-1067504391. Materia theme designers could have their own valid reason why they made focus color that way.

Synfig's unlucky to have a series of buttons packed in a row. Confused me a lot when I thought I was drawing both outlines and regions only to find out outlines was toggled off but the button's focus color deceived my eyes LOL.

> because we are breaking GTK behavior.

Not really. API for grab/ungrab focus is there in case application developers need them.

But this issue is debatable. Weighing in opinions from both sides (users & developers) would be much better.
We should keep both this issue and submitted PR open just in case.

**Comment by Keyikedalube** [comment]
> toolbar doesn't grab focus by a mouse click only by keyboard navigation

GTK toolbar buttons by design do not grab focus when clicked. That way the main application work area doesn't lose focus.

Imagine a text editor with toolbar buttons that can grab focus. Once a user clicks any button in the toolbar, editor's area focus is lost, he/she can no longer write or type stuff.

**Comment by mohamedAdhamc** [comment]
> > toolbar doesn't grab focus by a mouse click only by keyboard navigation
> 
> GTK toolbar buttons by design do not grab focus when clicked. That way the main application work area doesn't lose focus.
> 
> Imagine a text editor with toolbar buttons that can grab focus. Once a user clicks any button in the toolbar, editor's area focus is lost, he/she can no longer write or type stuff.

Yes, I meant just as focus by mouse click isn't needed there, I don't see a reason that it should be needed here. I might be missing a point here though.

**Comment by ice0** [comment]
> GTK toolbar buttons by design do not grab focus when clicked. That way the main application work area doesn't lose focus.
> 
> Imagine a text editor with toolbar buttons that can grab focus. Once a user clicks any button in the toolbar, editor's area focus is lost, he/she can no longer write or type stuff.

Agree, it makes sense.
Maybe it's just our problem, because we don't use `Gtk::Toolbar` for GTK4 compatibility, but we're still on GTK3.
